### DISCUSSION TITLE
Parameters and metadata from external file

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#7149] [Q&A]
### Description

**index.qmd**

````
---
title: '`r params$dataset`'
format: html
params:
  dataset: "iris"
---

This is the {{< meta params.dataset >}} dataset.

```{r,results="asis",echo=FALSE,comment=""}
cat(paste("This is the",params$dataset),"dataset")
```

```{r}
head(get(params$dataset))
```

````

**params.yml**

```
dataset: "mtcars"
```

`quarto render index.qmd --execute-params params.yml`

Two issues:

1. params need to be defined in the index.qmd for render to work. Seems extra work to define parameters twice
2. `{{< meta params.dataset >}}` is not replaced correctly.

![Screenshot 2023-10-05 at 21 51 19](https://github.com/quarto-dev/quarto-cli/assets/5083500/f9451d67-779a-4610-8a73-fab8db6a5c04)

```
▸ quarto --version
1.3.433
```

**Comment by mcanouil** [comment]
Firstly, you should use yaml style option as recommended in Quarto to avoid weird behaviour.

> params need to be defined in the index.qmd for render to work. Seems extra work to define parameters twice

This is on `knitr` not Quarto.

> `{{< meta params.dataset >}}` is not replaced correctly.

I don't think this is an error but simply a matter of execution order.
You should not use the meta shortcode to read the params which are not really metadata.

Example below:

```bash
quarto preview index.qmd --execute-params params.yml --metadata random-value:hello
```

```yml
dataset: "mtcars"
```

````qmd
---
format: html
params:
  dataset: "iris"
random-value: test
---

This is actual metadata: {{< meta random-value >}}

```{r}
#| output: asis
#| echo: false
cat(sprintf("---\ntitle: '%s'\n---\n", params$dataset))
```

This is the `{r} params$dataset` dataset.

```{r}
#| output: asis
#| echo: false
#| comment: ""
cat(paste("This is the", params$dataset), "dataset")
```

```{r}
head(get(params$dataset))
```
````

<img width="847" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/c0a80d87-116a-4f29-ba6a-a188c981d3b4">


Note: I am using the latest development version of Quarto which explains the "not usual" inline R expression (you need to remove the brackets in 1.3).

**Reply by royfrancis** [reply]
Thanks! This was helpful. But, is it possible to access params in **params.yml** through shortcode? I see that `meta` is not for params. I tried `var` but that doesn't work. 

I am a bit unsure why there is distinction between variables and metadata? They are both just yaml parameters. It should be possible to use them in text context or code context?

**Reply by mcanouil** [reply]
The issue is the order of execution.
Shortcodes are evaluated before computation and I believe passing the parameters is done afterwards.
Maybe this could be changed and done earlier which would then only the shortcode to read updated yaml frontmatter.
Not sure it is possible but I am not very familiar with that part of the codebase.
I'll have to take a deeper look (or let the other answer), but there might be an enhancement in here.

**Comment by royfrancis** [comment]
So, at the moment, if I want both of these to work:

```
This is the {{< meta dataset >}} dataset.
This is the `r params$dataset` dataset.
```
I should have two separate yaml matter:

```
params:
  dataset: iris
dataset: mtcars
```

And if I want them to be external files, params go to **params.yml** and used like `--execute-params params.yml`. What about the regular metadata? What format would that be and how do I pass it to the render function?

I have tried `--metadata metadata.yml`, but that doesn't seem to work.

**Reply by mcanouil** [reply]
As stated in the help, `--metadata` "- Metadata value (KEY:VALUE)." thus you can't use a file.
The correct option/command is `--metadata-file=FILE`.

Because you are using knitr, you can use `rmarkdown::metadata$dataset`.


````qmd
---
format: html
params:
  dataset: "iris"
dataset: mtcars
engine: knitr
---

- This is actual metadata: {{< meta dataset >}}
- This is the `{r} params$dataset` dataset.
- This is the `{r} rmarkdown::metadata$dataset` dataset.
````

**Reply by royfrancis** [reply]
Thanks! I couldn't find `--metadata-file` in `quarto render --help`.

**Reply by mcanouil** [reply]
`--metadata-file` as `--metadata` are actually Pandoc options.
Not all options are described in Quarto CLI help, but they do all work (in theory).

```
quarto render --help # a bit outdated though
quarto pandoc --help
```

**Comment by royfrancis** [comment]
Is it possible to use code in external metadata?

**index.qmd**

````
---
title: Test
format: html
params:
  dataset: iris
---

## {{< meta section >}}

{{< meta block >}}

This is the `r params$dataset` dataset.

```{r}
head(get(params$dataset))
```
````

**metadata.yml**

```
section: "Section A"
block: "We are analysing the `r params$dataset` dataset."
```

**params.yml**

```
dataset: "mtcars"
```

`quarto render index.qmd --execute-params params.yml --metadata-file metadata.yml`

![Screenshot 2023-10-06 at 11 28 52](https://github.com/quarto-dev/quarto-cli/assets/5083500/c1e50558-afbb-459b-9ad9-42646891f271)

Based on, shortcodes evaluated first and then code, this should work.

**Reply by cscheid** [reply]
> Based on, shortcodes evaluated first and then code, this should work.

That's not how it works. Shortcodes (except for include) are evaluated in Lua, inside Pandoc. knitr is executed before Pandoc.

knitr executes inline statements in the document, but (in almost all cases) not in the metadata. I don't think what you're trying to do is possible, unfortunately.

**Reply by mcanouil** [reply]
The question is: when does `--execute-params params.yml` happen? Should shortcode be aware of that file/parameter?

**Reply by cscheid** [reply]
> The question is: when does --execute-params params.yml happen? Should shortcode be aware of that file/parameter?

It can't. Anything that has to do with execution (including `--execute-params`) happens in the engines, before Pandoc. Shortcode evaluation (except for include) happens in Pandoc.

knitr doesn't (and shouldn't!) evaluate inline statements in metadata because it can't know in general if keys are meant as strings or markdown (and no, before someone who knows about `!expr` asks, we won't support `!expr` in there.) So that statement gets passed as text to Pandoc, and no evaluation of the inline statement happens.

**Reply by mcanouil** [reply]
> It can't. Anything that has to do with execution (including `--execute-params`) happens in the engines, before Pandoc. Shortcode evaluation (except for include) happens in Pandoc.

Thanks for confirming this.

> knitr doesn't (and shouldn't!) evaluate inline statements in metadata because it can't know in general if keys are meant as strings or markdown (and no, before someone who knows about `!expr` asks, we won't support `!expr` in there.) So that statement gets passed as text to Pandoc, and no evaluation of the inline statement happens.

I agree with this part (*i.e.* "shouldn't")!
